### ISSUE TITLE
DSL.multiset() should return a new org.jooq.Multiset type, which contains mapping convenience

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13539]
A few great `MULTISET` mapping utilities have been added to the jOOQ-kotlin module as extension functions:

- https://github.com/jOOQ/jOOQ/issues/12945
- https://github.com/jOOQ/jOOQ/issues/13538

However, if we had an `org.jooq.Multiset<R>` subtype of `Field<Result<R>>`, then we could attach these utilities directly to that type, and make them available as well to Java developers. For example, this would be possible:

```java
// Convenience ...
Field<Map<Col1Type, Col2Type>> field = multiset(select(T.COL1, T.COL2)).intoMap();

// ... for this
Field<Map<Col1Type, Col2Type>> field = multiset(select(T.COL1, T.COL2)).convertFrom(r -> r.map(Records.intoMap()));
```

A few additional methods (e.g. `Field<T>.as(...): Field<T>`) would have to be covariantly overridden as well.

**Comment by lukaseder** [comment]
The extension method approach is nicer, because it applies to *all* `Field<Result<R>>` structural types, irrespective how they come to be. This approach here using a nominal `Field<T>` subtype is not too nice, but we're already doing similar things with `Row[N]` types, which extend `SelectField<Record[N]>`

**Comment by lukaseder** [comment]
(I could have sworn there was already a feature request for `org.jooq.Multiset`, but I cannot find it)

**Comment by lukaseder** [comment]
Yeah, as expected, the ugliness of using subtypes to emulate better structural typing manifests with `MULTISET_AGG`, which exposes the aggregate (and window) function DSL API. Any DSL API step of `MULTISET_AGG` would have to be subtyped as well for these convenience methods to be available on `MULTISET_AGG`.

That's overkill and terrible design. I've rejected this approach early on in jOOQ, when I considered subtyping `Field<T>` for types of `T` (e.g. `IntegerField`, etc.). QueryDSL implemented this design, and I saw flaws everywhere. The blog post is here:
https://blog.jooq.org/the-dangers-of-correlating-subtype-polymorphism-with-generic-polymorphism/

`Row[N]` is slightly different, because:

1. It subtypes `SelectField<Record[N]>`, not `Field<Record[N]>`, which isn't as ubiquitous
2. It can't otherwise capture the `[N]` arity of the `Row[N]` and `Record[N]` types

I'll reject this issue here for now. The rejection is now documented (possibly again?), showing that this is a bad idea.
